core/src/HIP/Kokkos_HIP_Atomic.hpp

```diff
@@ -0,0 +1,591 @@
+  (uint64_t)(compare), (const uint64_t&)(val));
+  }
+
+  KOKKOS_INLINE_FUNCTION
+  float atomic_compare_exchange(float* dest, float compare, const float& val) {
+    union U {
+      int i ;
+      float f ;
+      KOKKOS_INLINE_FUNCTION U() {};
+    } idest,icompare,ival;
+    idest.f = *dest;
+    icompare.f = compare;
+    ival.f = val;
+    idest.i = hc::atomic_compare_exchange_int(reinterpret_cast<int*>(dest),
+  icompare.i, ival.i); return idest.f;
+  }
+
+  KOKKOS_INLINE_FUNCTION
+  double atomic_compare_exchange(double* dest, double compare, const double&
+  val) { union U { uint64_t i ; double d ; KOKKOS_INLINE_FUNCTION U() {}; }
+  idest,icompare,ival; idest.d = *dest; icompare.d = compare; ival.d = val;
+    idest.i =
+  hc::atomic_compare_exchange_uint64(reinterpret_cast<uint64_t*>(dest),
+  icompare.i, ival.i); return idest.d;
+  }
+
+  template<class T>
+  KOKKOS_INLINE_FUNCTION
+  T atomic_compare_exchange(volatile T* dest, T compare, typename
+  std::enable_if<sizeof(T) == sizeof(int), const T&>::type val) { union U { int
+  i ; T f ; KOKKOS_INLINE_FUNCTION U() {}; } idest,icompare,ival; idest.f =
+  *dest; icompare.f = compare; ival.f = val; idest.i =
+  hc::atomic_compare_exchange_int((int*)(dest), icompare.i, ival.i); return
+  idest.f;
+  }
+
+  template<class T>
+  KOKKOS_INLINE_FUNCTION
+  T atomic_compare_exchange(volatile T* dest, T compare, typename
+  std::enable_if<sizeof(T) == sizeof(int64_t), const T&>::type val) { union U {
+      uint64_t i ;
+      T f ;
+      KOKKOS_INLINE_FUNCTION U() {};
+    } idest,icompare,ival;
+    idest.f = *dest;
+    icompare.f = compare;
+    ival.f = val;
+    idest.i = hc::atomic_compare_exchange_uint64((uint64_t*)(dest), icompare.i,
+  ival.i); return idest.f;
+  }
+
+  template<class T>
+  KOKKOS_INLINE_FUNCTION
+  T atomic_compare_exchange(volatile T* dest, T compare, typename
+  std::enable_if<(sizeof(T) != sizeof(int32_t)) && (sizeof(T) !=
+  sizeof(int64_t)), const T&>::type val) { return val;
+  }
+*/
+inline __device__ int atomic_fetch_add(volatile int *dest, const int &val) {
+  return atomicAdd(const_cast<int *>(dest), val);
+}
+
+inline __device__ unsigned int atomic_fetch_add(volatile unsigned int *dest,
+                                                const unsigned int &val) {
+  return atomicAdd(const_cast<unsigned int *>(dest), val);
+}
+
+inline __device__ unsigned long long atomic_fetch_add(
+    volatile unsigned long long *dest, const unsigned long long &val) {
+  return atomicAdd(const_cast<unsigned long long *>(dest), val);
+}
+
+template <typename T>
+inline __device__ T atomic_fetch_add(volatile T *dest, const T &val) {
+  return val;
+}
+
+// FIXME Not implemented in HIP
+
+// KOKKOS_INLINE_FUNCTION
+// int64_t atomic_fetch_add(volatile int64_t* dest, const int64_t& val) {
+//   return atomicAdd(const_cast<int64_t*>(dest),val);
+// }
+KOKKOS_INLINE_FUNCTION
+char atomic_fetch_add(volatile char *dest, const char &val) {
```

Comment:
Why are all the other ones `__device__` and this one is KOKKOS_INLINE_FUNCTION? It still calls atomicCas internally (i.e. the device only function)